simpeg/directives/_regularization.py

```diff
@@ -0,0 +1,427 @@
+
+            # Reset the regularization matrices so that it is
+            # recalculated for current model. Do it to all levels of comboObj
+            for reg in self.reg.objfcts:
+                if not isinstance(reg, Sparse):
+                    continue
+
+                reg.update_weights(reg.model)
+
+            self.invProb.phi_m_last = self.reg(self.invProb.model)
+
+        # Repeat beta cooling schedule mechanism
+        if self.opt.iter > 0 and self.opt.iter % self.coolingRate == 0:
+            self.invProb.beta /= self.coolingFactor
+
+    def start_irls(self):
+        if self.verbose:
+            print(
+                "Reached starting chifact with l2-norm regularization:"
+                + " Start IRLS steps..."
+            )
+
+        if getattr(self.opt, "iter", None) is None:
+            self.metrics.start_irls_iter = 0
+        else:
+            self.metrics.start_irls_iter = self.opt.iter
+
+        self.invProb.phi_m_last = self.reg(self.invProb.model)
+
+        # Either use the supplied irls_threshold, or fix base on distribution of
+        # model values
+        for reg, norms in zip(self.reg.objfcts, self.metrics.input_norms):
+            if not isinstance(reg, Sparse):
+                continue
+
+            for obj in reg.objfcts:
+                threshold = np.percentile(
+                    np.abs(obj.mapping * obj._delta_m(self.invProb.model)),
+                    self.percentile,
+                )
+                if isinstance(obj, SmoothnessFirstOrder):
+                    threshold /= reg.regularization_mesh.base_length
+
+                obj.irls_threshold = threshold
+
+            reg.norms = norms
+
+            if self.verbose:
+                print("irls_threshold " + str(reg.objfcts[0].irls_threshold))
+
+        # Save l2-model
+        self.invProb.l2model = self.invProb.model.copy()
+
+    def validate(self, directiveList=None):
+        directive_list = directiveList.dList
+        self_ind = directive_list.index(self)
+        lin_precond_ind = [isinstance(d, UpdatePreconditioner) for d in directive_list]
+
+        if any(lin_precond_ind):
+            assert lin_precond_ind.index(True) > self_ind, (
+                "The directive 'UpdatePreconditioner' must be after Update_IRLS "
+                "in the directiveList"
+            )
+        else:
+            warnings.warn(
+                "Without a Linear preconditioner, convergence may be slow. "
+                "Consider adding `Directives.UpdatePreconditioner` to your "
+                "directives list",
+                stacklevel=2,
+            )
+
+        beta_schedule = [
+            d for d in directive_list if isinstance(d, BetaSchedule) and d != self
+        ]
+
+        assert len(beta_schedule) == 0, (
+            "Beta scheduling is handled by the `UpdateIRLS` directive."
+            "Remove the redundant `BetaSchedule` from your list of directives.",
+        )
```

Comment:
We could replace these assertion errors for explicit error raising. Also, the check for an empty list could be simplified to `if not beta_schedule`.
```suggestion
        if not beta_schedule:
            raise TypeError(
            "Beta scheduling is handled by the `UpdateIRLS` directive."
            "Remove the redundant `BetaSchedule` from your list of directives.",
        )
```